File: packages/guards/src/violence-detect.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ViolenceDetectOptions {
  action: 'block' | 'warn';
  categories?: ('physical' | 'weapon' | 'threat' | 'gore' | 'animal-abuse')[];
}

const VIOLENCE_KEYWORDS: Record<string, string[]> = {
  physical: ['punch', 'kick', 'stab', 'beat up', 'attack', 'assault', 'choke', 'strangle', 'murder', 'slaughter', 'massacre', 'torture', 'mutilate'],
  weapon: ['gun', 'knife', 'bomb', 'explosive', 'grenade', 'firearm', 'sword', 'machete', 'sniper', 'AK-47', 'AR-15', 'rifle', 'shotgun', 'ammunition'],
  threat: ['kill you', 'hurt you', 'destroy you', 'find where you live', 'come after you', 'watch your back', 'you will pay', 'death threat', 'i will end you'],
  gore: ['blood everywhere', 'dismember', 'decapitate', 'disembowel', 'severed', 'mutilated body', 'entrails', 'gore'],
  'animal-abuse': ['kick the dog', 'hurt the cat', 'animal cruelty', 'abuse animals', 'torture animals', 'kill the pet'],
};

const ALL_CATS = Object.keys(VIOLENCE_KEYWORDS);

export function violenceDetect(options: ViolenceDetectOptions): Guard {
  const cats = options.categories ?? ALL_CATS;
  return {
    name: 'violence-detect', version: '0.1.0',
    description: 'Detect violent content (physical, weapons, threats, gore)',
    category: 'content', supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const lower = text.toLowerCase();
      const triggered: Record<string, string[]> = {};
      for (const cat of cats) {
        const kws = VIOLENCE_KEYWORDS[cat] ?? [];
        const found = kws.filter((kw) => lower.includes(kw));
        if (found.length > 0) triggered[cat] = found;
      }
      const hasMatch = Object.keys(triggered).length > 0;
      return {
        guardName: 'violence-detect', passed: !hasMatch,
        action: hasMatch ? options.action : 'allow',
        message: hasMatch ? `Violence detected: ${Object.keys(triggered).join(', ')}` : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: hasMatch ? { triggered, reason: 'Text contains violent content' } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/violence_detect.py
"""Detects violence-related content in text."""
from __future__ import annotations

import re
import time
from typing import List

from open_guardrail.core import GuardResult

_PATTERNS = [
    re.compile(r"\bkill\b", re.IGNORECASE),
    re.compile(r"\bmurder\b", re.IGNORECASE),
    re.compile(r"\bassault\b", re.IGNORECASE),
    re.compile(r"\bweapon\b", re.IGNORECASE),
    re.compile(r"\bshoot\b", re.IGNORECASE),
    re.compile(r"\bstab\b", re.IGNORECASE),
    re.compile(r"\bbomb\b", re.IGNORECASE),
    re.compile(r"\battack\b", re.IGNORECASE),
]


class _ViolenceDetect:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "violence-detect"
        self.action = action

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matched: List[str] = []
        for p in _PATTERNS:
            m = p.search(text)
            if m:
                matched.append(m.group().lower())
        triggered = len(matched) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="violence-detect",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"Violence detected: {', '.join(matched)}" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"matched": matched} if triggered else None,
        )


def violence_detect(*, action: str = "block") -> _ViolenceDetect:
    return _ViolenceDetect(action=action)
